File: frontend/src/types/transaction.ts
export type TransactionType = "income" | "expense" | "transfer";

export interface Transaction {
  id: string;
  description: string;
  amount: number;
  type: TransactionType;
  date: string;
  user_id: string;
  category_id: string | null;
  account_id: string | null;
  transfer_id: string | null;
  transfer_direction: "in" | "out" | null;
  is_recurring: boolean;
  is_paid: boolean;
  installment_group_id: string | null;
  installment_number: number | null;
  installment_total: number | null;
}

export type CreateTransactionInput = {
  description: string;
  amount: number;
  type: TransactionType;
  date: string;
  category_id?: string | null;
  account_id?: string | null;
  is_recurring?: boolean;
  installments?: number;
};

export type UpdateTransactionInput = Partial<CreateTransactionInput>;

export interface TransferInput {
  from_account_id: string;
  to_account_id: string;
  amount: number;
  date: string;
  description?: string;
}

File: backend/src/schemas/transaction.py
from pydantic import BaseModel
from typing import Literal


TransactionType = Literal["income", "expense", "transfer"]


class TransactionBase(BaseModel):
    description: str
    amount: float
    type: TransactionType
    date: str
    category_id: str | None = None
    account_id: str | None = None
    is_recurring: bool = False


class CreateTransactionInput(TransactionBase):
    pass


class UpdateTransactionInput(BaseModel):
    description: str | None = None
    amount: float | None = None
    type: TransactionType | None = None
    date: str | None = None
    category_id: str | None = None
    account_id: str | None = None
    is_recurring: bool | None = None


class TransferInput(BaseModel):
    from_account_id: str
    to_account_id: str
    amount: float
    date: str
    description: str = "Transferência entre contas"


class Transaction(TransactionBase):
    id: str
    user_id: str
    transfer_id: str | None = None
    transfer_direction: str | None = None  # "out" | "in"

    model_config = {"from_attributes": True}


class TransactionBase(BaseModel):
    description: str
    amount: float
    type: TransactionType
    date: str
    category_id: str | None = None
    account_id: str | None = None
    is_recurring: bool = False
    is_paid: bool = True


class CreateTransactionInput(TransactionBase):
    # Campos de parcelamento — só usados quando installments > 1
    installments: int = 1


class UpdateTransactionInput(BaseModel):
    description: str | None = None
    amount: float | None = None
    type: TransactionType | None = None
    date: str | None = None
    category_id: str | None = None
    account_id: str | None = None
    is_recurring: bool | None = None
    is_paid: bool | None = None


class Transaction(TransactionBase):
    id: str
    user_id: str
    transfer_id: str | None = None
    transfer_direction: str | None = None
    installment_group_id: str | None = None
    installment_number: int | None = None
    installment_total: int | None = None

    model_config = {"from_attributes": True}
